core/solver/gcr.cpp

```diff
@@ -0,0 +1,358 @@
+    GKO_SOLVER_VECTOR(A_precon_residual, dense_b);
+    auto krylov_bases_p = this->create_workspace_op_with_type_of(
+        ws::krylov_bases_p, dense_b,
+        dim<2>{num_rows * (krylov_dim + 1), num_rhs});
+    auto mapped_krylov_bases_Ap = this->create_workspace_op_with_type_of(
+        ws::mapped_krylov_bases_Ap, dense_b,
+        dim<2>{num_rows * (krylov_dim + 1), num_rhs});
+    auto tmp_alpha = this->template create_workspace_op<Vector>(
+        ws::tmp_alpha, dim<2>{1, num_rhs});
+    auto tmp_beta = this->template create_workspace_op<Vector>(
+        ws::tmp_beta, dim<2>{1, num_rhs});
+    auto residual_norm = this->template create_workspace_op<NormVector>(
+        ws::residual_norm, dim<2>{1, num_rhs});
+    auto Ap_norms = this->template create_workspace_op<NormVector>(
+        ws::Ap_norms, dim<2>{krylov_dim + 1, num_rhs});
+    auto& final_iter_nums = this->template create_workspace_array<size_type>(
+        ws::final_iter_nums, num_rhs);
+
+    // indicates if one vectors status changed
+    GKO_SOLVER_ONE_MINUS_ONE();
+    bool one_changed{};
+    GKO_SOLVER_STOP_REDUCTION_ARRAYS();
+
+    // Initialization
+    // residual = dense_b
+    // reset stop status
+    exec->run(gcr::make_initialize(dense_b, residual, stop_status.get_data()));
+    // residual = residual - Ax
+    // Note: x is passed in with initial guess
+    this->get_system_matrix()->apply(neg_one_op, dense_x, one_op, residual);
+    this->get_preconditioner()->apply(residual, precon_residual);
+    // A_precon_residual = A*precon_residual
+    this->get_system_matrix()->apply(precon_residual, A_precon_residual);
+
+    // p(:, 1) = precon_residual(:, 1)
+    // Ap(:, 1) = A_precon_residual(:, 1)
+    // final_iter_nums = {0, ..., 0}
+    // apply preconditioner to residual
+    exec->run(gcr::make_restart(precon_residual, A_precon_residual,
+                                krylov_bases_p, mapped_krylov_bases_Ap,
+                                final_iter_nums.get_data()));
+
+    auto stop_criterion = this->get_stop_criterion_factory()->generate(
+        this->get_system_matrix(),
+        std::shared_ptr<const LinOp>(dense_b, [](const LinOp*) {}), dense_x,
+        residual);
+
+    int total_iter = -1;
+    size_type restart_iter = 0;
+
+    /* Memory movement summary for average iteration with krylov_dim d:
+     * (5d+25+6/d)n * values + matrix/preconditioner stroage
+     * 1x SpMV:                       2n * values + storage
+     * 1x Preconditioner:             2n * values + storage
+     * MGS:                   (5d + 21)n = sum k=0 to d-1 of (10k+26)n/d
+     *       1x dots             2(k+1)n in iteration k (0-based)
+     *       1x scals            2(k+1)n in iteration k (0-based)
+     *       2x axpys            6(k+1)n in iteration k (0-based)
+     *       1x norm2                  n
+     *       1x dot                   2n
+     *       1x sq_norm2               n
+     *       1x step 1 (scal, axpys)  8n
```

Comment:
should it be 6n?

Reply:
a += tmp * b // counted as 3n

Reply:
Step 1 has one dvide (x = a / b) and two axpy (a += tmp * b) so I counted 8n